python-lib/robust_session.py

```diff
@@ -0,0 +1,116 @@
+import logging
+import time
+
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO,
+                    format='sharepoint-online plugin %(levelname)s - %(message)s')
+
+
+class RobustSessionError(ValueError):
+    pass
+
+
+class RobustSession():
+    """
+    Implements a retry on status code 429 and connections reset by peer, and a connection reset + retry on error 403
+    """
+    def __init__(self, session=None, status_codes_to_retry=None, max_retries=1, base_retry_timer_sec=60):
+        logger.info("Init RobustSession")
+        self.session = session
+        self.status_codes_to_retry = status_codes_to_retry or []
+        self.max_retries = max_retries
+        self.base_retry_timer_sec = base_retry_timer_sec
+        self.connection_args = []
+        self.connection_kwargs = {}
+        self.connection_library = None
+
+    def update_settings(self, session=None, status_codes_to_retry=None, max_retries=None, base_retry_timer_sec=None):
+        self.session = session or self.session
+        self.status_codes_to_retry = status_codes_to_retry or self.status_codes_to_retry
+        self.max_retries = max_retries or self.max_retries
+        self.base_retry_timer_sec = base_retry_timer_sec or self.base_retry_timer_sec
+
+    def connect(self, connection_library=None, *args, **kwargs):
+        self.connection_library = connection_library or self.connection_library
+        self.connection_args = args or self.connection_args
+        self.connection_kwargs = kwargs or self.connection_kwargs
+        if self.connection_library:
+            self.session = self.retry(self.connection_library.connect, *self.connection_args, **self.connection_kwargs)
+
+    def get(self, url, dku_rs_off=False, **kwargs):
+        if dku_rs_off:
+            return self.session.get(url, **kwargs)
+        else:
+            kwargs["url"] = url
+            response = self.request_with_403_retry("get", **kwargs)
+            return response
+
+    def post(self, url, dku_rs_off=False, **kwargs):
+        if dku_rs_off:
+            return self.session.post(url, **kwargs)
+        else:
+            kwargs["url"] = url
+            response = self.request_with_403_retry("post", **kwargs)
+            return response
+
+    def request_with_403_retry(self, verb, **kwargs):
+        """
+        403 error code may be result of throttling, rendering the current sessions useless.
+        Therefore we try reset the session max_retries times before giving up.
+        """
+        attempt_number = 0
+        successful_request = False
+        while (not successful_request) and (attempt_number <= self.max_retries):
+            attempt_number += 1
+            if verb == "get":
+                response = self.retry(self.session.get, **kwargs)
+            else:
+                response = self.retry(self.session.post, **kwargs)
+            if response.status_code == 403:
```

Comment:
Out of curiosity, why does Microsoft reject our calls? Do they have a policy on how often their API can be called?